R/bundleFiles.R

```diff
@@ -0,0 +1,261 @@
+# build the list of files to deploy -- implicitly (directory contents),
+# explicitly via list, or explicitly via manifest. Always returns paths
+# related to `appDir`
+standardizeAppFiles <- function(appDir,
+                                appFiles = NULL,
+                                appFileManifest = NULL,
+                                error_call = caller_env()) {
+
+
+  no_content <- function(message) {
+    cli::cli_abort(
+      c("No content to deploy.", x = message),
+      call = error_call
+    )
+  }
+
+  if (!is.null(appFiles) && !is.null(appFileManifest)) {
+    cli::cli_abort(
+      "Must specify at most one of {.arg appFiles} and {.arg appFileManifest}",
+      call = error_call
+    )
+  } else if (is.null(appFiles) && is.null(appFileManifest)) {
+    # no files supplied at all, just bundle the whole directory
+    appFiles <- bundleFiles(appDir)
+    if (length(appFiles) == 0) {
+      no_content("{.arg appDir} is empty.")
+    }
+  } else if (!is.null(appFiles)) {
+    check_character(appFiles, allow_null = TRUE, call = error_call)
+    appFiles <- explodeFiles(appDir, appFiles)
+    if (length(appFiles) == 0) {
+      no_content("{.arg appFiles} didn't match any files in {.arg appDir}.")
+    }
+  } else if (!is.null(appFileManifest)) {
+    check_file(appFileManifest, error_call = error_call)
+    appFiles <- readFileManifest(appFileManifest)
+    appFiles <- explodeFiles(appDir, appFiles)
+    if (length(appFiles) == 0) {
+      no_content("{.arg appFileManifest} contains no usable files.")
+    }
+  }
+
+  appFiles
+}
+
+readFileManifest <- function(appFileManifest, error_call = caller_env()) {
+  lines <- readLines(appFileManifest, warn = FALSE)
+
+  # remove empty/comment lines
+  lines <- lines[nzchar(lines)]
+  lines <- lines[!grepl("^#", lines)]
+  lines
+}
+
+# given a list of mixed files and directories, explodes the directories
+# recursively into their constituent files, and returns just a list of files
+explodeFiles <- function(dir, files) {
+  exploded <- character()
+
+  totalFiles <- 0
+  totalSize <- 0
+
+  for (f in files) {
+    target <- file.path(dir, f)
+    if (!file.exists(target)) {
+      # doesn't exist
```

Comment:
nit: This may be warning-worthy, since we are exploding a set of files provided to us. Unchanged from the previous implementation. Fine to leave as-is and we can revisit.